eventuate-core/src/main/scala/com/rbmhtechnology/eventuate/ReplicationEndpoint.scala

```diff
@@ -548,27 +550,44 @@ private class Replicator(target: ReplicationTarget, source: ReplicationSource, f
 }
 
 private object FailureDetector {
-  case object Tick
+  case object AvailabilityDetected
+  case class FailureDetected(cause: Throwable)
+  case class FailureDetectionLimitReached(counter: Int)
 }
 
 private class FailureDetector(sourceEndpointId: String, logName: String, failureDetectionLimit: FiniteDuration) extends Actor {
-  import FailureDetector._
   import ReplicationEndpoint._
+  import FailureDetector._
+  import context.dispatcher
 
-  val failureDetectionLimitMillis = failureDetectionLimit.toMillis
-  var lastTick: Long = 0L
+  private var counter: Int = 0
+  private var causes: Vector[Throwable] = Vector.empty
+  private var schedule: Cancellable = scheduleFailureDetectionLimitReached()
 
-  context.setReceiveTimeout(failureDetectionLimit)
+  private val failureDetectionLimitNanos = failureDetectionLimit.toNanos
+  private var lastReportedAvailability: Long = 0L
 
   def receive = {
-    case Tick =>
-      val currentTime = System.currentTimeMillis
-      val lastInterval = currentTime - lastTick
-      if (lastInterval >= failureDetectionLimitMillis) {
+    case AvailabilityDetected =>
+      val currentTime = System.nanoTime()
+      val lastInterval = currentTime - lastReportedAvailability
+      if (lastInterval >= failureDetectionLimitNanos) {
         context.system.eventStream.publish(Available(sourceEndpointId, logName))
-        lastTick = currentTime
+        lastReportedAvailability = currentTime
       }
-    case ReceiveTimeout =>
-      context.system.eventStream.publish(Unavailable(sourceEndpointId, logName))
+      schedule.cancel()
+      schedule = scheduleFailureDetectionLimitReached()
+      causes = Vector.empty
+    case FailureDetected(cause) =>
+      causes = causes :+ cause
```

Comment:
Alternatively we could publish `Unavailable` if the timespan between the first and the last failure in this list exceeds `failureDetectionLimit` (given that the list is emptied on each `AvailabilityDetected`). We would not need any scheduler then, right?